container-engine-lib/lib/backend_impls/docker/docker_manager/docker_manager.go

```diff
@@ -954,6 +956,100 @@ func (manager *DockerManager) RunExecCommand(context context.Context, containerI
+	dockerClient := manager.dockerClient
+	execConfig := types.ExecConfig{
+		User:         "",
+		Privileged:   false,
+		Tty:          false,
+		AttachStdin:  false,
+		AttachStderr: true,
+		AttachStdout: true,
+		Detach:       false,
+		DetachKeys:   "",
+		Env:          nil,
+		WorkingDir:   "",
+		Cmd:          command,
+	}
+
+	createResp, err := dockerClient.ContainerExecCreate(context, containerId, execConfig)
+	if err != nil {
+		return nil, nil, stacktrace.Propagate(err, "An error occurred creating the exec process")
+	}
+
+	execId := createResp.ID
+	if execId == "" {
+		return nil, nil, stacktrace.NewError("Got back an empty exec ID when running '%v' on container '%v'", command, containerId)
+	}
+
+	execStartConfig := types.ExecStartCheck{
+		// Because detach is false, we'll block until the command comes back
+		Detach: false,
+		Tty:    false,
+	}
+
+	execOutputChan := make(chan string)
+	finalExecResultChan := make(chan *exec_result.ExecResult)
+	go func() {
+		defer func() {
+			close(execOutputChan)
+			close(finalExecResultChan)
+		}()
+
+		// IMPORTANT NOTE:
+		// You'd think that we'd need to call ContainerExecStart separately after this ContainerExecAttach....
+		//  ...but ContainerExecAttach **actually starts the exec command!!!!**
+		// We used to be doing them both, but then we were hitting this occasional race condition: https://github.com/moby/moby/issues/42408
+		// Therefore, we ONLY call Attach, without Start
+		attachResp, err := dockerClient.ContainerExecAttach(context, execId, execStartConfig)
+		if err != nil {
+			sendErrorThroughChannel(execOutputChan, err)
+			return
+		}
+		defer attachResp.Close()
+
+		// Stream output from docker through output channel
+		reader := bufio.NewReader(attachResp.Reader)
+		for {
+			execOutputLine, err := reader.ReadString(streamOutputDelimiter)
+			if err != nil {
+				if err == io.EOF {
+					break
+				} else {
+					sendErrorThroughChannel(execOutputChan, err)
+					return
+				}
+			}
+
+			execOutputChan <- execOutputLine
+		}
+
+		inspectResponse, err := dockerClient.ContainerExecInspect(context, execId)
+		if err != nil {
+			sendErrorThroughChannel(execOutputChan, err)
+			return
+		}
+		if inspectResponse.Running {
+			sendErrorThroughChannel(execOutputChan, stacktrace.NewError("Expected exec to have stopped, but it's still running!"))
+			return
+		}
+		unsizedExitCode := inspectResponse.ExitCode
+		if unsizedExitCode > math.MaxInt32 || unsizedExitCode < math.MinInt32 {
+			sendErrorThroughChannel(execOutputChan, stacktrace.NewError("Could not cast unsized int '%v' to int32 because it does not fit", unsizedExitCode))
+			return
+		}
+		int32ExitCode := int32(unsizedExitCode)
+
+		// Don't send output in final result because it was already streamed
+		finalExecResultChan <- exec_result.NewExecResult(int32ExitCode, "")
```

Comment:
Are we still using ExecResult for something else? Or should we just remove the string attribute from it entirely?

Reply:
It's used in the old `RunExecCommand` which I'm waiting to remove until the second PR for stream exec is up. But I'll note to remove it in the second PR!